lib/datadog/statsd/udp_connection.rb

```diff
@@ -19,18 +19,21 @@ def initialize(host, port, **kwargs)
 
         @host = host || ENV.fetch('DD_AGENT_HOST', DEFAULT_HOST)
         @port = port || ENV.fetch('DD_DOGSTATSD_PORT', DEFAULT_PORT).to_i
+        @socket = nil
+        connect
       end
 
       private
 
       def connect
-        UDPSocket.new.tap do |socket|
-          socket.connect(host, port)
-        end
+        @socket.close unless @socket == nil
+
+        @socket = UDPSocket.new
+        @socket.connect(host, port)
       end
```

Comment:
This change is ... weird in the context of `Connection#socket`. That method is 

```ruby
      def socket
        @socket ||= connect
      end
```

BUT because `connect` now manages the assignment, that `||=` will never actually do anything.

It looks to me that currently there's not even a code path where it ever gets executed for `UDPConnection`, but since `@socket.connect` does not return the `@socket`, if somehow `@socket == nil`  and `#socket` gets called, the `||=` will assign the result of `@socket.connect` to the `@socket`.

TL;DR, to avoid the complexity, I would suggest not assigning to the `@socket` directly OR to remove the `||=` from `#socket` and make it just an `attr_reader`.

Reply:
These magic getter (allocating if nil) are a bad idea overall. I'm removing it and will refactor the rest.

Reply:
Fixed in #204